File: ts/src/lib/grpc.ts
/**
 * A Yellowstone gRPC client that can see v1 transactions.
 *
 * `@triton-one/yellowstone-grpc` generates `Message.config` (field 7) from
 * 6.0.0 on. On 5.x the field is absent from the schema, and protobuf drops
 * unknown fields, so a v1 transaction decodes as v0 with no compute budget.
 */

import Client, {
    type ClientDuplexStream,
    CommitmentLevel,
    type SubscribeRequest,
    type SubscribeUpdate,
    type SubscribeUpdateTransactionInfo,
} from '@triton-one/yellowstone-grpc';

export type { SubscribeUpdate };

/**
 * A transaction message as the generated client decodes it.
 *
 * The package exports the `SubscribeUpdate` tree but not the `solana-storage`
 * types underneath it, so `Message` is reached through the update type.
 */
export type GrpcMessage = NonNullable<NonNullable<SubscribeUpdateTransactionInfo['transaction']>['message']>;

export type GrpcCompiledInstruction = GrpcMessage['instructions'][number];

export type MessageVersion = 'legacy' | 'v0' | 'v1';

/**
 * Identifies the version of a transaction message received over gRPC.
 *
 * The wire's `versioned` boolean is true for **both** v0 and v1. The only
 * signal is the presence of `config`, which nothing but v1 sets.
 */
export function messageVersion(message: GrpcMessage): MessageVersion {
    if (message.config !== undefined) {
        return 'v1';
    }
    return message.versioned ? 'v0' : 'legacy';
}

/** The empty filter maps every `SubscribeRequest` has to carry. */
const NO_FILTERS = {
    accounts: {},
    accountsDataSlice: [],
    blocks: {},
    blocksMeta: {},
    entry: {},
    slots: {},
    transactions: {},
    transactionsStatus: {},
} as const satisfies SubscribeRequest;

/** A subscription request for every non-vote, non-failed transaction. */
export const ALL_TRANSACTIONS_REQUEST: SubscribeRequest = {
    ...NO_FILTERS,
    commitment: CommitmentLevel.CONFIRMED,
    transactions: {
        all: { accountExclude: [], accountInclude: [], accountRequired: [], failed: false, vote: false },
    },
};

/** A subscription request for every block, with its transactions included. */
export const ALL_BLOCKS_REQUEST: SubscribeRequest = {
    ...NO_FILTERS,
    blocks: { all: { accountInclude: [], includeTransactions: true } },
    commitment: CommitmentLevel.CONFIRMED,
};

/**
 * Opens a subscription to a Yellowstone gRPC endpoint.
 *
 * `subscribe()` writes the request itself, so the caller does not resend it.
 * Destroying the stream releases the channel's keepalive timers, without which
 * the process would not exit on its own.
 */
export async function subscribe(
    endpoint: string,
    request: SubscribeRequest,
): Promise<{ close: () => void; stream: ClientDuplexStream }> {
    const client = new Client(endpoint, undefined, undefined);
    await client.connect();
    const stream = await client.subscribe(request);
    return { close: () => void stream.destroy(), stream };
}

export const DEFAULT_GRPC_ENDPOINT = process.env.TXV1_GRPC_URL ?? 'http://127.0.0.1:10000';

/**
 * Reads a non-negative integer limit from the environment, rejecting a typo
 * instead of ignoring it. Unset or empty means no limit.
 */
export function readEnvLimit(name: string): number {
    const raw = process.env[name];
    if (raw === undefined || raw === '') {
        return 0;
    }
    const value = Number(raw);
    if (!Number.isInteger(value) || value < 0) {
        throw new Error(`${name} must be a non-negative integer, got ${JSON.stringify(raw)}`);
    }
    return value;
}

/**
 * Exits the process when the subscription ends or fails.
 *
 * A production indexer would reconnect here and resume from the last processed
 * slot.
 */
export function exitOnStreamClose(stream: ClientDuplexStream, close: () => void): void {
    stream.on('end', () => {
        console.error('stream ended');
        close();
        process.exit(1);
    });

    stream.on('error', (error: Error) => {
        console.error(error.message);
        close();
        process.exit(1);
    });
}

File: rust/src/grpc.rs
//! Reading transaction versions and v1 configs off the Yellowstone gRPC wire.
//!
//! gRPC has no equivalent of `maxSupportedTransactionVersion`: no opt-in, no
//! version field, no server-side filter. Once the gate activates, v1 arrives
//! whether the consumer understands it or not.

use crate::MessageVersion;
use yellowstone_grpc_proto::solana::storage::confirmed_block::{Message, TransactionConfig};

/// Identifies the version of a transaction message received over gRPC.
///
/// The wire's `versioned` boolean is true for **both** v0 and v1. The only
/// signal is the presence of `config` (field 7), which nothing but v1 sets.
pub fn message_version(message: &Message) -> MessageVersion {
    match (&message.config, message.versioned) {
        (Some(_), _) => MessageVersion::V1,
        (None, true) => MessageVersion::V0,
        (None, false) => MessageVersion::Legacy,
    }
}

/// Renders a v1 `TransactionConfig` as indented lines.
///
/// An absent field is not a default: an unset limit resolves to zero, not to
/// the generous v0-era defaults.
pub fn describe_config(config: &TransactionConfig, indent: &str) -> String {
    let priority_fee = match config.priority_fee {
        Some(lamports) => format!("{lamports} lamports (total, not micro-lamports/CU)"),
        None => "unset (0)".to_string(),
    };
    [
        format!("{indent}priority_fee:                    {priority_fee}"),
        format!(
            "{indent}compute_unit_limit:              {}",
            optional_u32(config.compute_unit_limit, "0")
        ),
        format!(
            "{indent}loaded_accounts_data_size_limit: {}",
            optional_u32(config.loaded_accounts_data_size_limit, "0")
        ),
        format!(
            "{indent}heap_size:                       {}",
            optional_u32(config.heap_size, "32KB default")
        ),
    ]
    .join("\n")
}

fn optional_u32(value: Option<u32>, unset_meaning: &str) -> String {
    match value {
        Some(value) => value.to_string(),
        None => format!("unset ({unset_meaning})"),
    }
}
